Sources/HttpPipeline/NIO.swift

```diff
@@ -0,0 +1,135 @@
+import Foundation
+import NIO
+import NIOHTTP1
+import Optics
+import Prelude
+
+public func run(
+  _ middleware: @escaping Middleware<StatusLineOpen, ResponseEnded, Prelude.Unit, Data>,
+  on port: Int = 8080
+  ) {
+
+  do {
+    let reuseAddrOpt = ChannelOptions.socket(SocketOptionLevel(SOL_SOCKET), SO_REUSEADDR)
+    let group = MultiThreadedEventLoopGroup(numThreads: System.coreCount)
+    let bootstrap = ServerBootstrap(group: group)
+      .serverChannelOption(ChannelOptions.backlog, value: 256)
+      .serverChannelOption(reuseAddrOpt, value: 1)
+      .childChannelInitializer { channel in
+        channel.pipeline.configureHTTPServerPipeline().then {
+          channel.pipeline.add(handler: Handler(middleware))
+        }
+      }
+      .childChannelOption(ChannelOptions.socket(IPPROTO_TCP, TCP_NODELAY), value: 1)
+      .childChannelOption(reuseAddrOpt, value: 1)
+      .childChannelOption(ChannelOptions.maxMessagesPerRead, value: 1)
+
+    let serverChannel = try bootstrap.bind(host: "localhost", port: port).wait()
+    try serverChannel.closeFuture.wait()
+    try group.syncShutdownGracefully()
+  } catch {
+    fatalError(error.localizedDescription)
+  }
+}
+
+private final class Handler: ChannelInboundHandler {
+  typealias InboundIn = HTTPServerRequestPart
+
+  var request: URLRequest?
+  let middleware: Middleware<StatusLineOpen, ResponseEnded, Prelude.Unit, Data>
+
+  init(_ middleware: @escaping Middleware<StatusLineOpen, ResponseEnded, Prelude.Unit, Data>) {
+    self.middleware = middleware
+  }
+
+  func channelRead(ctx: ChannelHandlerContext, data: NIOAny) {
+    let reqPart = self.unwrapInboundIn(data)
+
+    switch reqPart {
+    case let .head(header):
+      self.request = URL(string: header.uri).map { url -> URLRequest in
+        var req = URLRequest(url: url)
+          |> \.httpMethod .~ method(from: header.method)
+          |> \.allHTTPHeaderFields .~ header.headers.reduce(into: [:]) { $0[$1.name] = $1.value }
+        req.url = req.value(forHTTPHeaderField: "Host")
+          .flatMap { URL(string: "http://" + $0 + url.absoluteString) }
```

Comment:
This logic is almost certainly wrong. Need to figure out how to properly construct the full request URL.